Sources/TensorFlow/Core/Serialization.swift

```diff
@@ -0,0 +1,85 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CTensorFlow
+import Foundation
+
+/// A TensorFlow checkpoint file reader.
+public class TensorFlowCheckpointReader {
+    @usableFromInline internal let status: OpaquePointer
+    @usableFromInline internal let handle: OpaquePointer
+
+    /// URL of the checkpoint file.
+    public let checkpointPath: URL
+
+    /// Number of tensors stored in the checkpoint.
+    public var tensorCount: Int { Int(TF_CheckpointReaderSize(handle)) }
+
+    /// Names of the tensors stored in the checkpoint.
+    public var tensorNames: [String] {
+        (0..<tensorCount).map {
+            String(cString: TF_CheckpointReaderGetVariable(handle, Int32($0)))
+        }
+    }
+
+    /// Creates a new TensorFlow checkpoint reader.
+    ///
+    /// - Arguments:
+    ///   - checkpointPath: URL of the checkpoint file.
+    @inlinable
+    public init?(checkpointPath: URL) {
+        self.status = TF_NewStatus()
+        self.handle = TF_NewCheckpointReader(checkpointPath.path, status)
+        checkOk(status)
+        self.checkpointPath = checkpointPath
+    }
+
+    deinit {
+        TF_DeleteCheckpointReader(handle)
+    }
+
+    /// Returns `true` if the checkpoint contains a tensor with the provided name.
+    @inlinable
+    public func contains(tensorNamed name: String) -> Bool {
+        TF_CheckpointReaderHasTensor(handle, name) > 0
+    }
+
+    /// Returns the shape of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func shape(ofTensorNamed name: String) -> TensorShape {
+        let rank = TF_CheckpointReaderGetVariableNumDims(handle, name)
+        let dimensions = UnsafeMutablePointer<Int64>.allocate(capacity: Int(rank))
+        defer { dimensions.deallocate() }
+        TF_CheckpointReaderGetVariableShape(handle, name, dimensions, rank, status)
+        checkOk(status)
+        let dimensionsBufferPointer = UnsafeBufferPointer(start: dimensions, count: Int(rank))
+        return TensorShape([Int64](dimensionsBufferPointer).map(Int.init))
+    }
+
+    /// Returns the data type of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func dataType(ofTensorNamed name: String) -> TensorDataType {
+        TensorDataType(TF_CheckpointReaderGetVariableDataType(handle, name))
+    }
+
+    /// Loads and returns the value of the tensor with the provided name stored in the checkpoint.
+    @inlinable
+    public func load<Scalar: _TensorFlowDataTypeCompatible>(
+        tensorNamed name: String
+    ) -> ShapedArray<Scalar> {
```

Comment:
I'm curious if there's some reason why `load` returns `ShapedArray<Scalar>` rather than `Tensor<Scalar>`?

Reply:
This is just because it is stored in host memory anyway and would require an internal conversion to Tensor. In this case, I thought it’d be nice to let the user decide to convert if they want to. What do you think?

Reply:
@dan-zheng do you think we can get this merged soon? I believe it is useful irrespective of how we design high-level serialization APIs as it allows us to read checkpoint files written by the TF Python API.

Reply:
> This is just because it is stored in host memory anyway and would require an internal conversion to Tensor. In this case, I thought it’d be nice to let the user decide to convert if they want to. What do you think?

I see, thanks for explaining! I'm curious if you use `ShapedArray` much in your projects, in addition to `Tensor`?

Reply:
Not really. The APIs for `ShapedArray` are very limited currently and I also mostly use GPUs for compute so I use `Tensor`s almost always. I just like to keep `ShapedArray` in APIs that could be used in both ways, so that it's up to the user to convert to a `Tensor`.